### DISCUSSION TITLE
Standalone stops working

### DISCUSSION THREAD

**Discussion by LexterS999** [0xERR0R/blocky#223] [Q&A]
Using standalone version and installed it on root folder with custom folder named blocky. Made setup using template on JSON example. And it's working fine but when I close SSH session console blocking ads stops working as Blocky too. How to make it work all the time? I even try to use my IP as main nameserver in hosts, but it didn't help. Any idea? Also, I don't understand how to use whitelist feature, and what format mus to be in whitelist.txt placed in folder?

**Comment by 0xERR0R** [comment]
How did you start blocky? It looks for me, like you just started `./blocky` from the shell and closed the shell. In this case, shell will kill all started child processes, this is correct. Try to start blocky with `nohup ./blocky` or use tmux. SystemD service would be a more preferable alternative (startup on restart etc)

**Comment by LexterS999** [comment]
> How did you start blocky? It looks for me, like you just started `./blocky` from the shell and closed the shell. In this case, shell will kill all started child processes, this is correct. Try to start blocky with `nohup ./blocky` or use tmux. SystemD service would be a more preferable alternative (startup on restart etc)

Ok, I will try. Is it ok to install it in custom folder or must be installed on root only?  And I started it with this cmd ./blocky --config config.yml

And what about this question: Also, I don't understand how to use whitelist feature, and what format mus to be in whitelist.txt placed in folder? How to make it right?

**Comment by LexterS999** [comment]
> SystemD service would be a more preferable alternative (startup on restart etc)

How to make that? I'm using Debian 9

**Comment by LexterS999** [comment]
When I'm trying to use this nohup ./blocky --config config.yml it's shows me next error: nohup: ignoring input and appending output to 'nohup.out'

Same error when I'm trying to use nohup ./blocky

**Comment by 0xERR0R** [comment]
> When I'm trying to use this nohup ./blocky --config config.yml it's shows me next error: nohup: ignoring input and appending output to 'nohup.out'
> 
> Same error when I'm trying to use nohup ./blocky

It is not an error, nohup creates a file "nohup.out", you can find blocky's output there (since you can now close you session and the process will run in the background)

**Comment by LexterS999** [comment]
Here is my config

upstream:
  default:
    - 1.1.1.1
    - 1.0.0.1
	- 8.8.8.8
	- 8.8.4.4
blocking:
  blackLists:
    ads:
      - https://raw.githubusercontent.com/StevenBlack/hosts/master/alternates/fakenews-gambling/hosts
	  - https://badmojr.github.io/1Hosts/Xtra/hosts.txt
	  - https://hosts.oisd.nl/
  whiteLists:
    ads:
      - whitelist.txt
  clientGroupsBlock:
    default:
      - ads
  blockType: zeroIp
  refreshPeriod: 240
  minTime: 5
  maxTime: 10
  prefetching: true
  port: 53
  disableIPv6: true

Showing next error
FATAL wrong file structure: yaml: line 5: found a tab character that violates indentation
FATAL wrong file structure: yaml: line 10: found a tab character that violates indentation
FATAL wrong file structure: yaml: line 11: found a tab character that violates indentation
  line 20: field minTime not found in type config.BlockingConfig
  line 21: field maxTime not found in type config.BlockingConfig
  line 22: field prefetching not found in type config.BlockingConfig
  line 23: field port not found in type config.BlockingConfig
  line 24: field disableIPv6 not found in type config.BlockingConfig

**Comment by LexterS999** [comment]
And what about this question: Also, I don't understand how to use whitelist feature, and what format mus to be in whitelist.txt placed in folder? How to make it right?

**Comment by 0xERR0R** [comment]
> Ok, I will try. Is it ok to install it in custom folder or must be installed on root only? And I started it with this cmd ./blocky --config config.yml

You can copy blocky to a random folder (/opt/blocky) for example.

> 
> And what about this question: Also, I don't understand how to use whitelist feature, and what format mus to be in whitelist.txt placed in folder? How to make it right?

Just create a text file with a domain per line

**Comment by 0xERR0R** [comment]
> Here is my config
> 
> upstream:
> default:
> - 1.1.1.1
> - 1.0.0.1
> - 8.8.8.8
> - 8.8.4.4
> blocking:
> blackLists:
> ads:
> - https://raw.githubusercontent.com/StevenBlack/hosts/master/alternates/fakenews-gambling/hosts
> - https://badmojr.github.io/1Hosts/Xtra/hosts.txt
> - https://hosts.oisd.nl/
> whiteLists:
> ads:
> - whitelist.txt
> clientGroupsBlock:
> default:
> - ads
> blockType: zeroIp
> refreshPeriod: 240
> minTime: 5
> maxTime: 10
> prefetching: true
> port: 53
> disableIPv6: true
> 
> Showing next error
> FATAL wrong file structure: yaml: line 5: found a tab character that violates indentation
> FATAL wrong file structure: yaml: line 10: found a tab character that violates indentation
> FATAL wrong file structure: yaml: line 11: found a tab character that violates indentation
> line 20: field minTime not found in type config.BlockingConfig
> line 21: field maxTime not found in type config.BlockingConfig
> line 22: field prefetching not found in type config.BlockingConfig
> line 23: field port not found in type config.BlockingConfig
> line 24: field disableIPv6 not found in type config.BlockingConfig

Config is in YAML format, indentation is very important. You can use this working example as template: https://raw.githubusercontent.com/0xERR0R/blocky/development/docs/config.yml

**Comment by LexterS999** [comment]
Yes, I know, and I used it as an example. So I showed you my config, but it's shows error all the time. What's wrong? Can you fix and provide right template based on my, please?

And I need to place whitelist.txt in the same folder where is placed blocky bin, right?

**Comment by LexterS999** [comment]
Here is my new template (attached as file
[template.txt](https://github.com/0xERR0R/blocky/files/6704829/template.txt)

But blocky still showing me next error:
[2021-06-23 22:57:54]  WARN net prefix udp is deprecated, using tcp+udp as default fallback
[2021-06-23 22:57:54]  WARN net prefix tcp is deprecated, using tcp+udp as default fallback
[2021-06-23 22:57:54] FATAL wrong file structure: yaml: unmarshal errors:
  line 51: field maxItemsCount not found in type config.CachingConfig
  line 58: field prefetchExpires not found in type config.CachingConfig
  line 61: field prefetchThreshold not found in type config.CachingConfig
  line 64: field prefetchMaxItemsCount not found in type config.CachingConfig

**Reply by 0xERR0R** [reply]
Are you using the latest version of blocky (0.14)?

**Comment by LexterS999** [comment]
Please help me and provide fixed config based on my template that I sent recent.
Also, if you can please provide how must be looked template for custom blocklist.txt and whitelist.txt

P/S And what format Blocky support? Domain, host, regex block?

**Reply by 0xERR0R** [reply]
Blocky supports the domain format (just one domain per line) or host format. Example for black/whitelist

```
www.youtube.com
www.youtu.be
```

**Comment by LexterS999** [comment]
Please help me and provide fixed config based on my template that I sent recently.

P/S As I suppose it's necessary to place a file like blacklist.txt inside a Blocky folder and paste domains, right? Same for whitelist?

**Comment by LexterS999** [comment]
How to apply this block configuration?

caching:
  # amount in minutes, how long a response must be cached (min value).
  # If <=0, use response's TTL, if >0 use this value, if TTL is smaller
  # Default: 0
  minTime: 5
  # amount in minutes, how long a response must be cached (max value).
  # If <0, do not cache responses
  # If 0, use TTL
  # If > 0, use this value, if TTL is greater
  # Default: 0
  maxTime: 10
  # Max number of cache entries (responses) to be kept in cache (soft limit). Useful on systems with limited amount of RAM.
  # Default (0): unlimited
  maxItemsCount: 3000000
  # if true, will preload DNS results for often used queries (default: names queried more than 5 times in a 2 hour time window)
  # this improves the response time for often used queries, but significantly increases external traffic
  # default: false
  prefetching: true
  # amount in minutes, prefetch track time window
  # default: 120
  prefetchExpires: 60
  # name queries threshold for prefetch
  # default: 5
  prefetchThreshold: 5
  # Max number of domains to be kept in cache for prefetching (soft limit). Useful on systems with limited amount of RAM.
  # Default (0): unlimited
  prefetchMaxItemsCount: 0

It's always saying showing next errors:

[2021-07-07 19:28:08] FATAL wrong file structure: yaml: unmarshal errors:
  line 23: field maxItemsCount not found in type config.CachingConfig
  line 30: field prefetchExpires not found in type config.CachingConfig
  line 33: field prefetchThreshold not found in type config.CachingConfig
  line 36: field prefetchMaxItemsCount not found in type config.CachingConfig


But I used config from example yml file. Can you help me and say what's wrong with it, please?

**Reply by 0xERR0R** [reply]
I assume, you're using not the latest version (v0.14). Please download 0.14 and try again

**Reply by LexterS999** [reply]
Yes, I'm already installed this version https://github.com/0xERR0R/blocky/releases/download/v0.14/blocky_0.14_Linux_x86_64.tar.gz 

But still something wrong with caching stings in configuration. And yes, I took config from the example which you provide. So something wrong in code, I guess? Any advice?

**Comment by 0xERR0R** [comment]
I just realized, that you're trying config from development branch, sorry, my bad ;(

Explaination: Development branch contains new feature of the future release (0.15). Master branch contains the last released version (0.14). maxItemsCount , prefetchExpires, etc.  are a part of the new feature. You can use the template of config.yaml from master: https://raw.githubusercontent.com/0xERR0R/blocky/master/docs/config.yml or readable version of configuration from official docs: https://0xerr0r.github.io/blocky/configuration/

**Reply by LexterS999** [reply]
Yes I used config from master branch but still the same error. So what's wrong with it? But I'm sure that installed  this version https://github.com/0xERR0R/blocky/releases/download/v0.14/blocky_0.14_Linux_x86_64.tar.gz

And how to install 0.15 version with new features and new config? Cause I need that features too

  # amount in minutes, prefetch track time window
  # default: 120
  prefetchExpires: 120
  # name queries threshold for prefetch
  # default: 5
  prefetchThreshold: 5
  # Max number of domains to be kept in cache for prefetching (soft limit). Useful on systems with limited amount of RAM.
  # Default (0): unlimited
  prefetchMaxItemsCount: 0

**Reply by 0xERR0R** [reply]
Just tried on a fresh linux machine:

```bash
mkdir blocky
cd blocky/
wget https://raw.githubusercontent.com/0xERR0R/blocky/master/docs/config.yml
wget https://github.com/0xERR0R/blocky/releases/download/v0.14/blocky_0.14_Linux_x86_64.tar.gz
tar -xvzf blocky_0.14_Linux_x86_64.tar.gz
chmod +x blocky
./blocky
```

Blocky starts, but I get `query log directory '/logs' does not exist or is not writable` -> this is ok, since the directory for logs does not exist. Changed "querylog.dir" from "/logs" to "." -> blocky runs

**Reply by LexterS999** [reply]
Yes, I did that, and it's working fine now, thanks. But I need also additional features like this (which is on 0.15 version):
  maxItemsCount: 0

  prefetching: true

  prefetchExpires: 120

  prefetchThreshold: 5

  prefetchMaxItemsCount: 0

**Reply by 0xERR0R** [reply]
There is no GitHub Release for the development branch. You can either use docker image with tag "development" or build it from the source code (checkout git repo, runt "make build"), but you must install Go.

Release 1.15 will come soon (this month).

**Reply by LexterS999** [reply]
> There is no GitHub Release for the development branch. You can either use docker image with tag "development" or build it from the source code (checkout git repo, runt "make build"), but you must install Go.
> 
> Release 1.15 will come soon (this month).

Cool, thanks for your time and patience. Then I will wait for new version but till this time 0.14 not bad, especially with cloudflared as DoH upstream on my local VPS. I suppose that prefetch will take replies from DoH the same as from usual public DNS, right?

And here is my suggestion: what if to add option to make changes in whitelists permanent without necessary of waiting updates? Cause refresh list commend doesn't work for me at all (don't know why, error with Prometheus, but I didn't use it). And it's hard to restart the whole Blocky from the begging each time.

**Reply by 0xERR0R** [reply]
So you mean, if you edit some whitelist.txt, the change must be propagated immediately? This could be done for local file, but it wouldn't work for network storages and files referenced as URL. You can trigger refresh manually by using the blocky CLI (https://0xerr0r.github.io/blocky/interfaces/) as `./blocky lists refresh` or just as REST call (https://0xerr0r.github.io/blocky/swagger.html) if you want to automate it

**Reply by LexterS999** [reply]
> So you mean, if you edit some whitelist.txt, the change must be propagated immediately? This could be done for local file, but it wouldn't work for network storages and files referenced as URL. You can trigger refresh manually by using the blocky CLI (https://0xerr0r.github.io/blocky/interfaces/) as `./blocky lists refresh` or just as REST call (https://0xerr0r.github.io/blocky/swagger.html) if you want to automate it

Yes, I mean it. But when I'm trying to apply next command, it's show me this error:
[2021-07-08 14:59:50] FATAL can't executePost "http://localhost:4000/api/lists/refresh": dial tcp [::1]:4000: connect: connection refused

And I'm not using Prometheus at all. Also take a note that I'm using nohup to make it work all the time, like this: 
cd /root/Blocky
nohup ./blocky --config config.yml

**Reply by 0xERR0R** [reply]
Are you running blocky cli on the same machine where blocky server runs?

**Reply by LexterS999** [reply]
As I told, I'm running it as a service. Cause when I close ssh terminal it's stop working. But even if I try to apply next cmd as CLI still getting error:

cd /root/Blocky
./blocky lists refresh

Getting this error
[2021-07-08 14:59:50] FATAL can't executePost "http://localhost:4000/api/lists/refresh": dial tcp [::1]:4000: connect: connection refused

And yes, I'm using it on the same remote VPS

**Reply by 0xERR0R** [reply]
Default config contains: "httpPort: 4000", if you are on the same server, than localhost:4000 should run. Please check, if blocky is running. If you start it with nohup, it will continue to run, even if you close your SSH session. Check the output in nohup file.

**Reply by LexterS999** [reply]
Got it. Now I know how to make it work right. Need to apply first this one
cd /root/Blocky
nohup ./blocky --config config.yml

And then apply ./blocky lists refresh

As I suppose it's update manually not only blacklists but whitelists too? Changes will be permanent, right?

**Reply by 0xERR0R** [reply]
Yes, it will update black and whitelists. Blocky holds all entries in memory, this is just a reload into memory

**Comment by LexterS999** [comment]
And think about next to add next feature  ./blocky log queries to see what Blocky filtering in real time. Because it's hard all the time looking inside log into folder.

**Reply by 0xERR0R** [reply]
Its a good idea, but not easy to implement: basically it should not rely on the log files on the file system (performance overhead, etc), the ideal solution would be to store all queries in a database and here comes the complexity: which database (sqllite? mysql? postgres?), configuration options etc. And this all for one small feature... 
Another option would be to store only last X entries in memory, but it is also not so future proof solution...
I'm thinking about it since 2 months, but I'm not sure, what is the best solution.

You could for example just open another terminal window and run `tail -f logfile.csv`` -> this will give you a realtime view (of course not so comfortable)

Note: I'm trying to keep blocky as simple as possible and combine it with existing tools (like prometheus). This is also the reason, why blocky has no web interface. There are 2 good feature-rich DNS solutions: Pi-Hole and AdGuard Home. Both can show queries in real-time. @asterriya :Did you try these 2?
